src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -999,13 +1003,33 @@ HeapWord* ShenandoahHeap::allocate_memory(ShenandoahAllocRequest& req) {
     // other threads have already depleted the free storage. In this case, a better
     // strategy is to try again, as long as GC makes progress (or until at least
     // one full GC has completed).
+
     size_t original_count = shenandoah_policy()->full_gc_count();
-    while (result == nullptr
-        && (get_gc_no_progress_count() == 0 || original_count == shenandoah_policy()->full_gc_count())) {
+
+    // Stop retrying and return nullptr to cause OOMError exception if our allocation failed even after:
+    //   a) We experienced a GC that had good progress, or
+    //   b) We experienced at least one Full GC (whether or not it had good progress)
+    //
+    // TODO: Rather than require a Full GC before throwing OOMError, it might be more appropriate for handle_alloc_failure()
+    //       to trigger a concurrent GLOBAL GC, and throw OOMError if we cannot allocate even after GLOBAL GC has finished.
+    //       There is no "perfect" solution here:
+    //
+    //        1. As currently implemented, there may be a race between multiple allocating threads, both attempting
+    //           to allocate very large objects.  The first thread to retry its allocation might succeed and the second
+    //           thread to retry its allocation might fail (because the first thread consumed the newly available memory).
+    //           So the second thread experiences OOMError even through another GC would have reclaimed the memory it wanted
+    //           to allocate.
+    //        2. A GLOBAL GC won't necessarily reclaim all garbage.  Following a concurrent Generational GLOBAL GC, we may
+    //           need to perform multiple concurrent mixed evacuations in order to reclaim all of the dead memory identified
+    //           by the GLOBAL GC mark.  However, the first evacuation performed by the GLOBAL GC will normally reclaim
+    //           a significant amount of garbage (as guided by garbage first heuristic).  If this is not enough memory
+    //           to satisfy the pending allocation request, we are in "dire straits", and a fail-fast OOMError is probably
+    //           the better remediation than repeated attempts to allocate following repeated GC cycles.
+
+    while ((result == nullptr) && (original_count == shenandoah_policy()->full_gc_count())) {
```

Comment:
These are subtle issues.  I'm not sure how best to address.  Will go with consensus.  One other observation:
```
Exception in thread thread_name: java.lang.OutOfMemoryError: GC Overhead limit exceeded
    Cause: The detail message "GC overhead limit exceeded" indicates that the garbage collector is running all the time and Java program is making very slow progress. After a garbage collection, if the Java process is spending more than approximately 98% of its time doing garbage collection and if it is recovering less than 2% of the heap and has been doing so far the last 5 (compile time constant) consecutive garbage collections, then a java.lang.OutOfMemoryError is thrown. This exception is typically thrown because the amount of live data barely fits into the Java heap having little free space for new allocations.
```
from https://docs.oracle.com/javase/8/docs/technotes/guides/troubleshoot/memleaks002.html

This description aligns with our default behavior of ShenandoahNoProgressThreshold default of 5, which would trigger in this PR's current implementation if we do 5 back-to-back unproductive full GCs (regardless of how much CPU time is consumed by GC vs mutator threads).  Unproductive is defined as NOT is_good_progress(), where is_good_progress() (see shenandoahMetrics.cpp) is:

1. free memory is at least 1% of the heap capacity, and
    a. The amount of used memory is decreased by at least the size of one region, or
    b. We have decreased internal fragmentation by at least 1%, or
    c. We have decreased external fragmentation by at least 1%

One problem that I have observed with certain test programs is that this definition of "is_good_progress" is pretty generous.  We can have many back-to-back full gcs that "make progress" according to this generous definition, while a mutator thread repeatedly retries and fails to satisfy its allocation request.

My judgement is that the mutator is much better served by getting a quick OOM exception rather than chugging away in this retry loop.